drivers/include/periph/countdown.h

```diff
@@ -0,0 +1,261 @@
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    drivers_periph_countdown Countdown timer
+ * @ingroup     drivers_periph
+ * @brief       Low-level countdown peripheral driver
+ *
+ *              Countdown timers are cheap timers that count down from a set
+ *              number, then execute an alarm callback.
+ *              Focus is on speedy execution of the .set() / .start() functions
+ *              which allows them to be used in time critical situations.
+ *
+ *              Typically the .set() function will be more costly, so a timeout
+ *              can be configured in advance that then only needs to be started
+ *              by calling the light-weight .start() function.
+ * @{
+ * @file
+ * @brief       Low-level countdown peripheral driver interface definitions
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#ifndef PERIPH_COUNTDOWN_H
+#define PERIPH_COUNTDOWN_H
+
+#include <limits.h>
+#include <stdint.h>
+
+#include "timex.h"
+#include "periph_cpu.h"
+#include "periph_conf.h"
+#include "xfa.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Countdown API flags
+ * @{
+ */
+#define COUNTDOWN_FLAG_UNIT_NS  (0 << 0)    /**< time is in ns  */
+#define COUNTDOWN_FLAG_UNIT_US  (1 << 0)    /**< time is in µs  */
+#define COUNTDOWN_FLAG_UNIT_MS  (2 << 0)    /**< time is in ms  */
+#define COUNTDOWN_FLAG_UNIT_SEC (3 << 0)    /**< time is in s   */
+#define COUNTDOWN_FLAG_UNIT_MASK (0x3)      /**< time unit mask */
+
+#define COUNTDOWN_FLAG_PERIODIC (1 << 2)    /**< Countdown will restart when it
+                                                 reaches zero */
+/** @} */
+
+/**
+ * @brief Type of a countdown timer
+ */
+typedef unsigned countdown_t;
+
+/**
+ * @brief Countdown alarm callback function
+ *
+ * @param[in] arg   Optional user supplied function argument
+ */
+typedef void (*countdown_cb_t)(void *arg);
+
+/**
+ * @brief Countdown driver API implementation
+ */
+typedef struct countdown_desc countdown_desc_t;
+
+/**
+ * @brief   Countdown device
+ */
+typedef struct {
+    const countdown_desc_t *driver; /**< API implementation */
+    const void *ctx;                /**< driver context */
+} countdown_dev_t;
+
+/**
+ * @brief   Countdown device run-time config
+ */
+typedef struct {
+#if IS_USED(MODULE_PERIPH_COUNTDOWN_LONG)
+    uint32_t max_timeout;       /**< largest timeout period in base unit */
+    uint32_t skip_cb[2];        /**< number of max timeout periods remaining / total */
+    uint32_t remaining;         /**< remaining ticks after max timeout periods */
+    countdown_cb_t real_cb;     /**< original user callback */
+    void *real_ctx;             /**< original user callback argument */
+#endif
+    uint32_t ticks_per_unit;    /**< counter ticks per base unit */
+    uint8_t base_unit;          /**< base unit (ns, µs, …) of the counter */
```

Comment:
My idea was that you typically don't (or shouldn't) need to care about the timer frequency.
So you only tell `countdown_init()` about the time range you want to have for the countdown and then it figures out the best frequency on it's own (where `ticks_per_unit` would be a whole number).

I *think* it was then easier to calculate the actual timeouts based on that (since you can just multiply with `ticks_per_unit`).

It's been a while since I while since I wrote this, might have to take a look at the API again to see if it still makes sense.